numpy/linalg/linalg.py

```diff
@@ -2788,10 +2795,11 @@ def _multi_dot_matrix_chain_order(arrays, return_costs=False):
     return (s, m) if return_costs else s
 
 
-def _multi_dot(arrays, order, i, j):
+def _multi_dot(arrays, order, i, j, out=None):
     """Actually do the multiplication with the given order."""
     if i == j:
         return arrays[i]
     else:
         return dot(_multi_dot(arrays, order, i, order[i, j]),
-                   _multi_dot(arrays, order, order[i, j] + 1, j))
+                   _multi_dot(arrays, order, order[i, j] + 1, j),
```

Comment:
Should the recursive calls to `_multi_dot` get the `out` argument? What about in the `i == j` case?

Reply:
`out` should be used only by the final call to `np.dot`. The intermediate products before the final product might be a different shape from `out`. So only the first call to `_multi_dot` from `multi_dot` should get `out`, and the recursive `_multi_dot` calls will just get `out=None` and behave as before

`i == j` is only used as the terminating case of the recursive `_multi_dot` calls. The initial call to `_multi_dot` from `multi_dot` is `_multi_dot(arrays, order, 0, n - 1, out=out)`, so it will never have `i == j` because `n` cannot equal 1 at that point. The first call will always go down the `np.dot` code path and use the `out` for the final output

Reply:
A very brief comment next to `return arrays[i]` to that effect would be helpful to future readers.

Reply:
added in 80c3034cff020f4c6af542f6c9f7a3bcea598319